flang/module/__fortran_builtin_kinds.f90

```diff
@@ -0,0 +1,110 @@
+!===-- module/__fortran_bultin_kinds.f90 --=--------------------------------===!
```

Comment:
why can't these definitions remain in iso_fortran_env?

Reply:
Flang was giving me compiler errors for numeric_storage_size, which seems it needs to not be compiled as part of the module file but rather as part of when the module has been included into the host routine.

That's I basically went back to my original idea of how I wanted to fix the bug and moved the intrinsic definitions into their own module file.

Reply:
__builtin_numeric_storage_size() gets special treatment in folding that can easily be adjusted without this kind of disruption.

Reply:
Can you provide a hint how this could be done?  I'm not that deep into the inner workings for Flang yet to see how this can be done.

Reply:
Grep for `builtin_numeric_storage` and you'll see what's going on.

Reply:
I have done that already.  Thing is: when the context is right, `__build_numeric_storage_size` is immediately folded.  However, then compiling `iso_fortran_env.f90`, line 1324-1326 in `fold-integer.cpp` simply leave the evaluation of `__builtin_numeric_storage_size` in the intermediate code, so that the code-gen picks it up and utterly complains about not having code-gen for it.  So, I have tried to add that, but that does not seem the right path to me.

Reply:
Ping :-)

Reply:
yes?

Reply:
Is my sketched approach the correct one or is there a simpler and better way to keep the original expression when compiling the module file?

Reply:
I'm confident that the problem can be solved without splitting the module source file by properly defining the compilations in which the folding the intrinsic must be deferred.

Reply:
Honestly, I did not see how.  So, hints or even a patch about how to do this will be appreciated.

Reply:
In `flang/lib/Evaluate/fold-integer.cpp` ca. line 1340 is the code that folds (underscore underscore) `builtin_numeric_storage`.  It defers that folding today until the reference to that intrinsic function is coming from a module file, at which point the size of a default numeric storage unit must be known.  If you're changing things so that `ISO_FORTRAN_ENV` is now going to be compiled with the size of a default numeric storage unit being well-defined, then you can probably just remove the code that defers the folding.

Reply:
When I remove the code, the compiler folds the current value `32_4` into the `iso_fortran_env.mod` file.  From there on, `numeric_storage_size` always reports 32 as the storage size, regardless of compiler options.

The code removed:

```C++
    if (!context.moduleFileName()) {
      // Don't fold this reference until it appears in the module file
      // for ISO_FORTRAN_ENV -- the value depends on the compiler options
      // that might be in force.
    } else {
```

Reply:
Yes, but since you now need to also compile the modules into distinct `*.o` files for each combination of target and relevant compilation options, isn't that what has to happen?

Reply:
That's not needed, if we do the separation like I'm proposing in this PR.  Everything that is compiled into an object file is independent of the relevant compiler options (like `-fdefault-real-8`), while `numeric_storage_size` is left in `iso_fortran_env` that is only treated with `-fsyntax-only`, which resolves the whole issue w/o touching the compiler itself.

Reply:
And complicated module file shenanigans are better somehow than "touching" the compiler?  I guess that's a matter of opinion.  I'll let you find some other reviewer for this, but you can help them out by at least documenting what's going on here with some comments, and fixing the name of the file in its header.